rfcs/2023-09-19-18604-metrics-data-model-refactor.md

```diff
@@ -0,0 +1,455 @@
+such as taking in a fixed-bucket histogram from the `prometheus_scrape` source and being able to get
+a DDSketch to send out from the `datadog_metrics` sink.
+
+Said another way, the goal here is to be able to hand a `Distribution` metric to any metrics-capable
+sink and allow it to represent that distribution in the highest fidelity way possible, whether
+that's losslessly converting from one distribution variant to another, or emitting fixed quantiles
+from a sketch, and so on.
+
+#### Raw distribution
+
+The `Raw` variant would match the current shape of `Distribution` where the raw samples are stored, which
+includes the sample value and the sample rate. This variant is used for sources where the sample
+value comes in directly (such as StatsD histograms) or use cases like the `log_to_metric` transform,
+where the value is also coming directly from a log event.
+
+#### Histograms
+
+The `Histogram` variant would be the spiritual successor to `AggregatedHistogram`, where we have a true
+histogram (bucketed values) but the number of buckets, and the bucket sizes, are fixed. While a
+DDSketch can be decomposed into its individual buckets, such that we could theoretically use a
+DDSketch to emit a Prometheus-capable aggregated histogram, we have no mechanism to store the
+original buckets used from the source metric within a DDSketch. Since Prometheus' own aggregation
+depends on having equivalent buckets for aggregated histograms, we're providing this variant to
+handle that use case.
+
+All of that said, this variant can still be remapped into a DDSketch (via interpolation) and so it
+is otherwise compatible with the goal of a `Distribution` being able to represent itself in multiple
+forms while still allowing for conversion to the highest fidelity represent for a given consumer of
+that distribution metric.
+
+#### Sketches
+
+This variant would be the merged version of the original `Sketch` variant, which currently only
+contains the `DDSketch` variant, mapped to the DDSketch configuration parameters used by the Datadog
+Agent.
+
+Similarly to aggregated histograms, this variant specifically supports the Datadog Agent use case,
+which sends distributions as a compressed/condensed data structure called `DDSketch` that allows for
+summarizing the distribution data in a space-efficient way while still providing a bound on the
+error of values estimated by the sketch.
+
+While we could additionally refactor this to only ever refer to the agent-specific configuration of
+DDSketch -- as we do not currently support any other sketch implementations/configurations of
+DDSketch -- we're leaving this as-is because the current sketch code is fairly complex and is likely
+subject to change in the future. Simplification can be done at that time, and the enum-based
+approach currently used for the particular sketch variant provides us an escape hatch for that
+future refactoring work.
```

Comment:
Do you foresee us leveraging the fact that it's an enum in some future change? E.g. introducing a `DDSketchV2` or similar as part of a refactoring? I'm mostly curious if you foresee the API changing or just the internals, because it would be nice to collapse this additional level of conceptual indirection while we're doing this modeling change.

Reply:
If I had to lean a particular way, I'd say it would be an API change.

If the agent, for example, ever changed to use a version of DDSketch that was closer to the way DDSketch is implemented in the original paper, it would be an API-level change (changing field types inside `DDSketch`, things that couldn't be configured without type-erased generics) and that's where we'd definitely need to add an enum variant.

I do agree that the current situation is not great, but perhaps we could split the difference and just hoist up `MetricSketch` into `Sketch` itself?

Reply:
I'd be in favor of anything that reduces the chain you have to follow from `Distribution::Sketch -> MetricSketch -> MetricSketch::AgentDDSketch -> AgentDDSketch`. I don't know how likely it is that the Agent changes its sketch implementation, or how much it would buy us to already have the enum in that case, but there is definitely a carrying cost to the more complex modeling in the meantime. I'm fine with whatever we think is the right balance, I just want to make sure we're thinking about it and simplifying where we can.

Reply:
It would definitely be reasonable for us to just have `enum Distribution { Sketch(AgentDDSketch) }` given, to your point, that it is not particularly likely that the Agent changes its sketch representation. Even if it did, we'd have a correspondingly reasonable time to adjust to that if need be since it would be a very big change for them to make and handle the rollout of.

Reply:
Addressed in 7b92f3684.

Now we have `Distribution::Sketch { sketch: AgentDDSketch }` which just gets rid of the indirection of `MetricSketch` entirely.